src/pybind/mgr/dashboard/services/service.py

```diff
@@ -0,0 +1,164 @@
+from .orchestrator import OrchClient
+
+logger = logging.getLogger('service')
+
+
+class NoCredentialsException(Exception):
+    def __init__(self):
+        super(NoCredentialsException, self).__init__(
+            'No RGW credentials found, '
+            'please consult the documentation on how to enable RGW for '
+            'the dashboard.')
+
+
+class RgwServiceManager:
+    def restart_rgw_daemons_and_set_credentials(self):
+        # Restart RGW daemons and set credentials.
+        orch = OrchClient.instance()
+        services, _ = orch.services.list(service_type='rgw', offset=0)
+
+        all_daemons_up = True
+        for service in services:
+            daemons_up = self.verify_service_restart('rgw', service['service_id'])
+            if not daemons_up:
+                all_daemons_up = False
+
+        if all_daemons_up:
+            self.configure_rgw_credentials()
+
+    def verify_service_restart(self, type: str, id: str):
+        orch = OrchClient.instance()
+        service_name = f'{type}.{id}'
+
+        # Get initial service info
+        info = orch.services.get(service_name)[0].to_dict()
+        last_refreshed = info['status']['last_refresh']
+
+        # Reload the service
+        orch.services.reload(type, id)
+
+        # Wait for the service to refresh
+        self.wait_for_refresh(orch, service_name, last_refreshed)
+
+        # Check the daemon status
+        daemon_status = self.wait_for_daemon_to_start(orch, service_name)
+        return daemon_status
+
+    def wait_for_refresh(self, orch, service_name, last_refreshed):
+        orch = OrchClient.instance()
+        while True:
+            updated_info = orch.services.get(service_name)[0].to_dict()
+            if updated_info['status']['last_refresh'] != last_refreshed:
+                break
+
+    def wait_for_daemon_to_start(self, orch, service_name):
+        orch = OrchClient.instance()
+        start_time = time.time()
+        while True:
+            daemons = [d.to_dict() for d in orch.services.list_daemons(service_name=service_name)]
+            all_running: bool = True
+
+            for daemon in daemons:
+                daemon_state = daemon['status_desc']
+
+                if daemon_state in ('unknown', 'error', 'stopped'):
+                    raise DashboardException(
+                        code='daemon_restart_failed',
+                        msg=f"Failed to restart the daemon {service_name}. Daemon state is {daemon_state}."  # noqa E501  #pylint: disable=line-too-long
+                    )
+                if daemon_state != 'running':
+                    all_running = False
+
+            if all_running:
+                return True
+
+            if time.time() - start_time > 10:
+                raise DashboardException(
+                    code='daemon_restart_timeout',
+                    msg=f"Timeout reached while waiting for daemon {service_name} to start."
+                )
+            return False
```

Comment:
these are generic to every service and not just rgw so it would be helpful in the future if this is moved to a more generic ServiceManager or as a standalone function outside this class..

Reply:
Done. Thanks